File: typescript/06-memoria/20-implementaciones/mem0_integration.ts
// Integración de Mem0 como capa de memoria para un agente existente.
//
// Mem0 extrae memorias automáticamente de cada turno conversacional
// mediante un LLM auxiliar. La integración mínima requiere 3 llamadas:
// add() al final de cada turno, search() al inicio, getAll() para contexto completo.
//
// Requiere:
//   export MEM0_API_KEY=...
//   export ANTHROPIC_API_KEY=...
//
// Cómo ejecutar:
//   export MEM0_API_KEY=tu-clave
//   export ANTHROPIC_API_KEY=tu-clave
//   make ts SCRIPT=typescript/06-memoria/20-implementaciones/mem0_integration.ts
//
// Qué esperar:
//   El agente extrae memorias automáticamente de cada turno. Muestra el ciclo
//   add() → search() → getAll() en acción.

import Anthropic from "@anthropic-ai/sdk";

const MEM0_BASE = "https://api.mem0.ai/v1";
const MEM0_API_KEY = process.env.MEM0_API_KEY ?? "";
const MODEL = process.env.MODEL ?? "claude-sonnet-4-6";
const USER_ID = "usuario-demo-ts";

interface Mem0Memory {
  id: string;
  memory: string;
}

function mem0Headers() {
  return {
    Authorization: `Token ${MEM0_API_KEY}`,
    "Content-Type": "application/json",
  };
}

async function mem0Add(messages: Array<{ role: string; content: string }>, userId: string): Promise<void> {
  const res = await fetch(`${MEM0_BASE}/memories/`, {
    method: "POST",
    headers: mem0Headers(),
    body: JSON.stringify({ messages, user_id: userId }),
  });
  if (!res.ok) throw new Error(`Mem0 add error ${res.status}: ${await res.text()}`);
}

async function mem0Search(query: string, userId: string, limit = 5): Promise<Mem0Memory[]> {
  const params = new URLSearchParams({ query, user_id: userId, limit: String(limit) });
  const res = await fetch(`${MEM0_BASE}/memories/search/?${params}`, {
    headers: { Authorization: `Token ${MEM0_API_KEY}` },
  });
  if (!res.ok) throw new Error(`Mem0 search error ${res.status}`);
  return (await res.json()) as Mem0Memory[];
}

async function mem0GetAll(userId: string): Promise<Mem0Memory[]> {
  const params = new URLSearchParams({ user_id: userId });
  const res = await fetch(`${MEM0_BASE}/memories/?${params}`, {
    headers: { Authorization: `Token ${MEM0_API_KEY}` },
  });
  if (!res.ok) throw new Error(`Mem0 getAll error ${res.status}`);
  return (await res.json()) as Mem0Memory[];
}

async function recuperarContexto(query: string): Promise<string> {
  const resultados = await mem0Search(query, USER_ID);
  if (!resultados.length) return "";
  const lineas = resultados.map((r) => `- ${r.memory}`);
  return "## Memoria recuperada\n" + lineas.join("\n");
}

async function turno(
  historial: Array<{ role: string; content: string }>,
  mensaje: string
): Promise<string> {
  const cliente = new Anthropic();

  // 1. Recuperar contexto relevante antes de responder
  const contextoMemoria = await recuperarContexto(mensaje);
  let system = "Eres un asistente técnico.";
  if (contextoMemoria) system += `\n\n${contextoMemoria}`;

  historial.push({ role: "user", content: mensaje });
  const respuestaApi = await cliente.messages.create({
    model: MODEL,
    max_tokens: 1024,
    system,
    messages: historial,
  });
  const respuesta = (respuestaApi.content[0] as { text: string }).text;
  historial.push({ role: "assistant", content: respuesta });

  // 2. Guardar el turno para sesiones futuras (post-turno — fuera del hot path ideal)
  await mem0Add([{ role: "user", content: mensaje }, { role: "assistant", content: respuesta }], USER_ID);
  return respuesta;
}

// ── main ──────────────────────────────────────────────────────────────────

if (!MEM0_API_KEY) {
  console.error("MEM0_API_KEY no configurada. Exporta la clave y reintenta.");
  process.exit(1);
}
if (!process.env.ANTHROPIC_API_KEY) {
  console.error("ANTHROPIC_API_KEY no configurada.");
  process.exit(1);
}

const historial: Array<{ role: string; content: string }> = [];

// Turno 1: el agente aprende la preferencia
const r1 = await turno(historial, "Prefiero trabajar con Python 3.12 en producción.");
console.log(`Agente: ${r1.slice(0, 120)}\n`);

// Turno 2: nueva sesión — historial vacío, pero Mem0 recupera la preferencia
const historialNuevo: Array<{ role: string; content: string }> = [];
const r2 = await turno(historialNuevo, "¿Qué lenguaje debería usar para el nuevo servicio?");
console.log(`Agente (sesión nueva): ${r2.slice(0, 120)}`);

// Ver todas las memorias guardadas
console.log("\n--- memorias almacenadas ---");
for (const m of await mem0GetAll(USER_ID)) {
  console.log(`  ${m.memory}`);
}

File: go/06-memoria/20-implementaciones/mem0_integration.go
// Integración de Mem0 como capa de memoria para un agente existente.
//
// Mem0 extrae memorias automáticamente de cada turno conversacional
// mediante un LLM auxiliar. La integración mínima requiere 3 llamadas:
// add() al final de cada turno, search() al inicio, getAll() para contexto completo.
//
// Requiere:
//
//	export MEM0_API_KEY=...
//	export ANTHROPIC_API_KEY=...
//
// Cómo ejecutar:
//
//	export MEM0_API_KEY=tu-clave
//	export ANTHROPIC_API_KEY=tu-clave
//	make go FILE=go/06-memoria/20-implementaciones/mem0_integration.go
//
// Qué esperar:
//
//	El agente extrae memorias automáticamente de cada turno. Muestra el ciclo
//	add() → search() → getAll() en acción.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
)

const (
	mem0Base = "https://api.mem0.ai/v1"
	apiBase  = "https://api.anthropic.com/v1"
	userID   = "usuario-demo-go"
)

var (
	mem0Key    = os.Getenv("MEM0_API_KEY")
	anthropicKey = os.Getenv("ANTHROPIC_API_KEY")
	modelo = func() string {
		if v := os.Getenv("MODEL"); v != "" {
			return v
		}
		return "claude-sonnet-4-6"
	}()
)

type Mem0Memory struct {
	ID     string `json:"id"`
	Memory string `json:"memory"`
}

func mem0Add(messages []map[string]string, uid string) error {
	body, _ := json.Marshal(map[string]any{"messages": messages, "user_id": uid})
	req, _ := http.NewRequest("POST", mem0Base+"/memories/", bytes.NewReader(body))
	req.Header.Set("Authorization", "Token "+mem0Key)
	req.Header.Set("Content-Type", "application/json")
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 300 {
		data, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("Mem0 add error %d: %s", resp.StatusCode, string(data[:minI(len(data), 200)]))
	}
	return nil
}

func mem0Search(query, uid string, limit int) ([]Mem0Memory, error) {
	params := url.Values{"query": {query}, "user_id": {uid}, "limit": {fmt.Sprint(limit)}}
	req, _ := http.NewRequest("GET", mem0Base+"/memories/search/?"+params.Encode(), nil)
	req.Header.Set("Authorization", "Token "+mem0Key)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var mems []Mem0Memory
	json.NewDecoder(resp.Body).Decode(&mems) //nolint:errcheck
	return mems, nil
}

func mem0GetAll(uid string) ([]Mem0Memory, error) {
	params := url.Values{"user_id": {uid}}
	req, _ := http.NewRequest("GET", mem0Base+"/memories/?"+params.Encode(), nil)
	req.Header.Set("Authorization", "Token "+mem0Key)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	var mems []Mem0Memory
	json.NewDecoder(resp.Body).Decode(&mems) //nolint:errcheck
	return mems, nil
}

func recuperarContexto(query string) string {
	mems, err := mem0Search(query, userID, 5)
	if err != nil || len(mems) == 0 {
		return ""
	}
	out := "## Memoria recuperada\n"
	for _, m := range mems {
		out += "- " + m.Memory + "\n"
	}
	return out
}

type Mensaje struct {
	Role    string `json:"role"`
	Content string `json:"content"`
}

func turno(historial []Mensaje, mensaje string) (string, []Mensaje, error) {
	// 1. Recuperar contexto relevante
	system := "Eres un asistente técnico."
	if ctx := recuperarContexto(mensaje); ctx != "" {
		system += "\n\n" + ctx
	}

	historial = append(historial, Mensaje{Role: "user", Content: mensaje})

	body, _ := json.Marshal(map[string]any{
		"model":      modelo,
		"max_tokens": 1024,
		"system":     system,
		"messages":   historial,
	})
	req, _ := http.NewRequest("POST", apiBase+"/messages", bytes.NewReader(body))
	req.Header.Set("x-api-key", anthropicKey)
	req.Header.Set("anthropic-version", "2023-06-01")
	req.Header.Set("Content-Type", "application/json")

	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		return "", historial, err
	}
	defer resp.Body.Close()

	var result struct {
		Content []struct {
			Text string `json:"text"`
		} `json:"content"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil || len(result.Content) == 0 {
		return "", historial, fmt.Errorf("error decodificando respuesta")
	}
	respuesta := result.Content[0].Text
	historial = append(historial, Mensaje{Role: "assistant", Content: respuesta})

	// 2. Guardar turno para sesiones futuras (post-turno — fuera del hot path ideal)
	mem0Add([]map[string]string{
		{"role": "user", "content": mensaje},
		{"role": "assistant", "content": respuesta},
	}, userID) //nolint:errcheck

	return respuesta, historial, nil
}

func main() {
	if mem0Key == "" {
		fmt.Fprintln(os.Stderr, "MEM0_API_KEY no configurada. Exporta la clave y reintenta.")
		os.Exit(1)
	}
	if anthropicKey == "" {
		fmt.Fprintln(os.Stderr, "ANTHROPIC_API_KEY no configurada.")
		os.Exit(1)
	}

	var historial []Mensaje

	// Turno 1: el agente aprende la preferencia
	r1, historial, err := turno(historial, "Prefiero trabajar con Python 3.12 en producción.")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error turno 1: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Agente: %s\n\n", r1[:minI(len(r1), 120)])

	// Turno 2: nueva sesión — historial vacío, pero Mem0 recupera la preferencia
	var historialNuevo []Mensaje
	r2, _, err := turno(historialNuevo, "¿Qué lenguaje debería usar para el nuevo servicio?")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error turno 2: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Agente (sesión nueva): %s\n", r2[:minI(len(r2), 120)])

	// Ver todas las memorias guardadas
	fmt.Println("\n--- memorias almacenadas ---")
	mems, _ := mem0GetAll(userID)
	for _, m := range mems {
		fmt.Printf("  %s\n", m.Memory)
	}

	_ = historial
}

func minI(a, b int) int {
	if a < b {
		return a
	}
	return b
}
